src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -0,0 +1,572 @@
+        JsonArray jsonTotpDevices = parseAndValidateFieldType(userData, "totpDevices", ValueType.ARRAY_OF_OBJECT, false,
+                JsonArray.class, errors, ".");
+
+        if (jsonTotpDevices == null) {
+            return null;
+        }
+
+        if (Arrays.stream(FeatureFlag.getInstance(main, appIdentifier).getEnabledFeatures())
+                .noneMatch(t -> t == EE_FEATURES.MFA)) {
+            errors.add("MFA must be enabled to import totp devices.");
+            return null;
+        }
+
+        List<TotpDevice> totpDevices = new ArrayList<>();
+        for (JsonElement jsonTotpDeviceEl : jsonTotpDevices) {
+            JsonObject jsonTotpDevice = jsonTotpDeviceEl.getAsJsonObject();
+
+            String secretKey = parseAndValidateFieldType(jsonTotpDevice, "secretKey", ValueType.STRING, true,
+                    String.class, errors, " for a totp device.");
+            Integer period = parseAndValidateFieldType(jsonTotpDevice, "period", ValueType.INTEGER, false,
+                    Integer.class, errors, " for a totp device.");
+            Integer skew = parseAndValidateFieldType(jsonTotpDevice, "skew", ValueType.INTEGER, false, Integer.class,
+                    errors, " for a totp device.");
+            String deviceName = parseAndValidateFieldType(jsonTotpDevice, "deviceName", ValueType.STRING, false,
+                    String.class, errors, " for a totp device.");
+
+            secretKey = validateAndNormaliseTotpSecretKey(secretKey, errors);
+            period = validateAndNormaliseTotpPeriod(period, errors);
+            skew = validateAndNormaliseTotpSkew(skew, errors);
+            deviceName = validateAndNormaliseTotpDeviceName(deviceName, errors);
+
+            if (secretKey != null && period != null && skew != null) {
+                totpDevices.add(new TotpDevice(secretKey, period, skew, deviceName));
+            }
+        }
+        return totpDevices;
+    }
+
+    private List<LoginMethod> getParsedLoginMethods(Main main, AppIdentifier appIdentifier, JsonObject userData,
+            List<String> errors)
+            throws StorageQueryException, TenantOrAppNotFoundException {
+        JsonArray jsonLoginMethods = parseAndValidateFieldType(userData, "loginMethods", ValueType.ARRAY_OF_OBJECT,
+                true, JsonArray.class, errors, ".");
+
+        if (jsonLoginMethods == null) {
+            return new ArrayList<>();
+        }
+
+        if (jsonLoginMethods.size() == 0) {
+            errors.add("At least one loginMethod is required.");
+            return new ArrayList<>();
+        }
+
+        if (jsonLoginMethods.size() > 1) {
+            if (Arrays.stream(FeatureFlag.getInstance(main, appIdentifier).getEnabledFeatures())
+                    .noneMatch(t -> t == EE_FEATURES.ACCOUNT_LINKING || t == EE_FEATURES.MFA)) {
+                errors.add("Account linking or MFA must be enabled to import multiple loginMethods.");
```

Comment:
find this code in AuthRecipe and extract a util function

Reply:
done